packages/slate-editor/src/modules/nodes-hierarchy/hierarchySchema.ts

```diff
@@ -32,71 +50,141 @@ import { IMAGE_CANDIDATE_NODE_TYPE } from '#extensions/image/constants';
+        allowChildren(
+            isEmptyTextNode,
+            combineFixers([unwrapSameTypeChild, liftNodeWithSplit, unwrapNode]),
+        ),
     ],
     [IMAGE_CANDIDATE_NODE_TYPE]: [
-        allowChildren(isEmptyTextNode, combineFixers([unwrapSameTypeChild, liftNode, unwrapNode])),
+        allowChildren(
+            isEmptyTextNode,
+            combineFixers([unwrapSameTypeChild, liftNodeWithSplit, unwrapNode]),
+        ),
     ],
     [IMAGE_NODE_TYPE]: [
-        allowChildren(isTextNode, combineFixers([unwrapSameTypeChild, liftNode, unwrapNode])),
+        allowChildren(
+            Text.isText,
+            combineFixers([unwrapSameTypeChild, liftNodeWithSplit, unwrapNode]),
+        ),
     ],
     [LOADER_NODE_TYPE]: [
-        allowChildren(isEmptyTextNode, combineFixers([unwrapSameTypeChild, liftNode, unwrapNode])),
+        allowChildren(
+            isEmptyTextNode,
+            combineFixers([unwrapSameTypeChild, liftNodeWithSplit, unwrapNode]),
+        ),
     ],
     [PARAGRAPH_NODE_TYPE]: [
-        allowChildren(isInlineNode, combineFixers([unwrapSameTypeChild, liftNode, unwrapNode])),
+        allowChildren(
+            isInlineNode,
+            combineFixers([unwrapSameTypeChild, liftNodeWithSplit, unwrapNode]),
+        ),
     ],
     [QUOTE_NODE_TYPE]: [
         allowChildren(
             isInlineNode,
             combineFixers([
                 (editor, [node, path]) => isParagraphNode(node) && unwrapNode(editor, [node, path]),
                 unwrapSameTypeChild,
-                liftNode,
+                liftNodeWithSplit,
                 unwrapNode,
             ]),
         ),
     ],
     [STORY_BOOKMARK_NODE_TYPE]: [
-        allowChildren(isEmptyTextNode, combineFixers([unwrapSameTypeChild, liftNode, unwrapNode])),
+        allowChildren(
+            isEmptyTextNode,
+            combineFixers([unwrapSameTypeChild, liftNodeWithSplit, unwrapNode]),
+        ),
     ],
     [STORY_EMBED_NODE_TYPE]: [
-        allowChildren(isEmptyTextNode, combineFixers([unwrapSameTypeChild, liftNode, unwrapNode])),
+        allowChildren(
+            isEmptyTextNode,
+            combineFixers([unwrapSameTypeChild, liftNodeWithSplit, unwrapNode]),
+        ),
+    ],
+    [TABLE_CELL_NODE_TYPE]: [
+        allowChildren(
+            isAllowedInTableCell,
+            combineFixers([unwrapSameTypeChild, liftNodeNoSplit, unwrapNode, convertToParagraph]),
+        ),
+    ],
+    [TABLE_NODE_TYPE]: [
+        allowChildren(
+            isTableRowNode,
+            combineFixers([unwrapSameTypeChild, liftNodeNoSplit, unwrapNode]),
+        ),
+    ],
+    [TABLE_ROW_NODE_TYPE]: [
+        allowChildren(
+            isTableCellNode,
+            combineFixers([unwrapSameTypeChild, liftNodeNoSplit, unwrapNode, convertToParagraph]),
+        ),
     ],
+    [TEXT_NODE_TYPE]: [isInside(isTableCellNode, optimizeTextInsideTableCell)],
```

Comment:
hm, i treat the first function as a condition and the function inside is a fixer
like here 
```
    [PARAGRAPH_NODE_TYPE]: [
        allowChildren(
            isInlineNode,
            combineFixers([unwrapSameTypeChild, liftNodeWithSplit, unwrapNode]),
        ),
    ],
 ```
 
 I think that `allowChildren` is condition and function inside is a fixer
 
 In this api 
 ```
 [TEXT_NODE_TYPE]: [
   disallowMark('bold', descendantsOf(isTableCellNode)),
],
```
`disallowMark` works as fixer because it would actually remove mark from node and `descendantsOf` is condition

Reply:
> I think that `allowChildren` is condition and function inside is a fixer

In contrast, I see `allowChildren` as a normalization *rule*. And then there should be a fixer provided as a parameter, as there's no single solution to all cases. Just to illustrate my vision, try to imagine `allowChildren` not requiring a fixer (it's optional; if omitted, it will utilize its built-in default normalization strategy). Then it'll look this way (which does make sense too):

```tsx
[...]: allowChildren(isParagraphNode),
```

`disallowMark` is a rule too. But because there's only one way to solve it (remove the mark) we don't need to pass a fixer to it.

IMO, normalization schema should consist of *rules*. So function names should read like rules. Consistency and order everywhere :nerd_face:

Reply:
> In contrast, I see `allowChildren` as a normalization rule. And then there should be a fixer provided as a parameter, as there's no single solution to all cases. Just to illustrate my vision, try to imagine allowChildren not requiring a fixer (it's optional; if omitted, it will utilize its built-in default normalization strategy). Then it'll look this way (which does make sense too)

We can also fix this confusion by dropping the `HierarchyFixer` interface and changing `allowChildren` to accept a normalizer as a second parameter (because they are semantically the same):

```tsx
export function allowChildren(
    isAllowed: HierarchyNodeQuery,
    normalize: HierarchyNormalizer,
) { ... }
```

Reply:
Hm, i see what you mean, just from my point of view `allowChildren` is just a condition and the normalizers themselves are inside. It looks like matter of naming, we can rename `allowChildren` to `whenChildren`
imo we have slate node type, condition, normalizer when condition is true

Reply:
Yes, you are right that the naming can be aligned in both directions.

I just genuinely believe that the *rules* semantics is a better way to go (see my example with `allowChildren` without fixers). It does make more sense, IMO.

--- 

Also, just by reading the code, I will definitely get what this will do:

```ts
disallowMark('bold', descendantsOf(isTableCellNode))
```

while this looks confusing:

```ts
isInside(isTableCellNode, optimizeTextInsideTableCell)
```